src/main/java/org/humancellatlas/ingest/project/ProjectService.java

```diff
@@ -152,7 +152,7 @@ public void delete(Project project) throws NonEmptyProject {
         ProjectBag projectBag = gather(project);
         if (projectBag.submissionEnvelopes.isEmpty()) {
             projectBag.projects.forEach(_project -> {
-                projectRepository.delete(_project);
+                metadataCrudService.deleteDocument(_project);
```

Comment:
I think that instead of calling:
```
ProjectService -> MetaDataCrudService -> Project Repository
```
and adding extra delete work in the CrudService, it would be possible and also potentially cleaner to use a `RepositoryEventHandler` that would listen to repository events.

This would mean to keep the call chain:

```
ProjectService -> ProjectRepository
```
and that our pal Spring would invoke the repository event handler that would take care of the extra operations required before or after the repo deletion.

Here is an example for the ProjectRepo. There would have to be a similar thing for the other repos.
```java
@RepositoryEventHandler(Project.class) 
public class ProjectEventHandler {
    private final @NonNull ProjectRepository projectRepository;
    private final @NonNull ProtocolRepository protocolRepository;
    //... and all the other dependencies added to ProjectCrudStrategy

    @HandleBeforeDelete
    public void beforeDelete(Project document){
        messageRouter.routeStateTrackingDeleteMessageFor(document);
        biomaterialRepository.findByProject(document).forEach(biomaterial -> {
            biomaterial.setProject(null);
            biomaterialRepository.save(biomaterial);
        });
        // ... and all the additional repository operations from ProjectCrudStrategy. removeLinksToDocument()
    }
}
```

Relevant docs:
- [example](https://www.baeldung.com/spring-data-rest-events)
- [spring docs](https://docs.spring.io/spring-data/rest/docs/current/reference/html/#events)

Not sure this is mandatory.